lib/include/tree_sitter/api.h

```diff
@@ -134,10 +134,30 @@ typedef enum {
   TSQueryErrorLanguage,
 } TSQueryError;
 
+typedef struct {
+  void *(*malloc)(size_t);
+  void *(*calloc)(size_t, size_t);
+  void *(*realloc)(void *, size_t);
+  void (*free)(void *);
+} TSAllocator;
+
 /********************/
 /* Section - Parser */
 /********************/
 
+/**
+ * Switch to a new allocator.
+ *
+ * Returns true iff the switch successes.
+ *
+ * This function can only be invoked *once*, before tree-sitter has allocated
+ * any memory via malloc/calloc/realloc. Otherwise, memory bugs could occur
+ * since some memory is allocated by the old allocator, but freed by the new
+ * one.
```

Comment:
I actually don't think we need to impose this "only once" restriction in the library. Applications can maintain this invariant themselves. There are valid use cases where the user may want to call this function more than once, and it should work provided they either:
1. Ensure all trees are freed before calling it
2. Provide an allocator that shares some state with the old allocator.

For instance, once we add this feature, I'd like to refactor the Tree-sitter test suite to use `ts_set_allocator` to implement the "allocation tracking" which is currently done using a compile time conditional. That should work fine, since the tracking allocator just *wraps* the normal allocator.

So could you just remove the logic for enforcing that this is only called once, and make this return `void`?

We could still include some kind of warning in the doc comment, but I also think it's reasonable to expect users to decide this for themselves.

Reply:
> Applications can maintain this invariant themselves. There are valid use cases where applications may want to call this function more than once, and it should work, as long as the application either:
> 
> 1. Ensures all parsers and trees are freed before calling it
> 2. Provides an allocator that shares its state with the old allocator.
> 

Can I use these comments (with slight modifications) as the doc comment before the function?


> So could you just remove the logic for enforcing that this is only called once, and make this return `void`?

Do we need to provide an interface to return current allocator functions?

Reply:
> > Applications can maintain this invariant themselves. There are valid use cases where applications may want to call this function more than once, and it should work, as long as the application either:
> > 
> > 1. Ensures all parsers and trees are freed before calling it
> > 2. Provides an allocator that shares its state with the old allocator.
> 
> Can I use these comments (with slight modifications) as the doc comment before the function?
> 

I added the two cases in the doc comment before the function. If you think I need to rephrase in my own words, please let me know.